beacon-chain/rpc/validator/proposer.go

```diff
@@ -194,6 +206,128 @@ func (vs *Server) eth1Data(ctx context.Context, slot uint64) (*ethpb.Eth1Data, e
 	return eth1Data, nil
 }
 
+// eth1DataMajorityVote determines the appropriate eth1data for a block proposal using an extended
+// simple voting algorithm - voting with the majority. The algorithm for this method is as follows:
+//  - Determine if we are in the "head" of the voting period. If yes, use the simple voting algorithm for proposal.
+//  - Determine the timestamp for the start slot for the current eth1 voting period.
+//  - Determine the timestamp for the start slot for the previous eth1 voting period.
+//  - Determine the most recent eth1 block before each timestamp.
+//  - Subtract the current period's eth1block.number by ETH1_FOLLOW_DISTANCE to determine the voting upper bound.
+//  - Subtract the previous period's eth1block.number by ETH1_FOLLOW_DISTANCE to determine the voting lower bound.
+//  - Filter out votes on unknown blocks and blocks which are outside of the range determined by the lower and upper bounds.
+//  - If no blocks are left after filtering, use the current period's most recent eth1 block for proposal.
+//  - Determine the vote with the highest count. Prefer the vote with the highest eth1 block height in the event of a tie.
+//  - This vote's block is the eth1block to use for the block proposal.
+// TODO: "Genesis" period and previous period
+// TODO: Feature flag
+func (vs *Server) eth1DataMajorityVote(ctx context.Context, slot uint64) (*ethpb.Eth1Data, error) {
```

Comment:
`eth1DataMajorityVote` is long, i suggest we refactor some into smaller functions. 
For one example, the aggregation logic can be it's own function

Reply:
+1, this function is doing 10 different things. It can be refactored into a few smaller functions

Reply:
+1 this method is hard to parse, its doing too many things and should be broken up in smaller functions